agora-test/Spec/Effect/TreasuryWithdrawal.hs

```diff
@@ -10,28 +10,272 @@ module Spec.Effect.TreasuryWithdrawal (tests) where
 import Agora.Effect.TreasuryWithdrawal (
   TreasuryWithdrawalDatum (TreasuryWithdrawalDatum),
   treasuryWithdrawalValidator,
+  PTreasuryWithdrawalDatum
  )
+import Plutus.V1.Ledger.Api
+import Plutarch.Builtin
 import Plutus.V1.Ledger.Value qualified as Value
-import Sample.Effect.TreasuryWithdrawal (
-  buildReceiversOutputFromDatum,
-  buildScriptContext,
-  currSymbol,
-  inputCollateral,
-  inputGAT,
-  inputTreasury,
-  inputUser,
-  outputTreasury,
-  outputUser,
-  treasuries,
-  users,
- )
-import Test.Tasty (TestTree, testGroup)
+import Plutus.V1.Ledger.Interval qualified as Interval
+import Plutarch.Api.V1
+import Sample.Effect.TreasuryWithdrawal
+import Sample.Sample
 import Test.Util (effectFailsWith, effectSucceedsWith)
+import Agora.Utils
+import Test.QuickCheck
+import Test.QuickCheck.Monadic
+
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.QuickCheck 
+import Test.Tasty.Plutarch.Property (classifiedProperty)
+
+import Data.Tagged
+import Data.Universe
+
+import Control.Applicative
+
+type TWETestInput = (TreasuryWithdrawalDatum, TxInfo)
+data TWETestCases
+  = PaysToEffect
+  | OutputsDoNotMatchReceivers
+  | InputsHaveOtherScriptInput
+  | RemaindersDoNotReturnToTreasuries
+  | EffectShouldPass
```

Comment:
This is quite useful. I'm curious, how do we know when we have covered all the cases? Specifically, how do we know that these four are the reasons scripts will fail?

Reply:
Currently it's using classified property and will be replaced with quite faster `alwaysFailProperty`. In either cases, tests will check generated incorrect case fails. 

You've mentioned having multiple/combined version; I think that is possible to implement(with PCB). However, coverage on specific cases ensures other permutations, so don't really think that's needed